File: output/CodeGen/codenet/C++/Python/s877835664.py
def main():
    N, K = map(int, raw_input().split())
    A = map(int, raw_input().split())
    dic = [-1] * (N + 1)
    cur = 1
    flg = False
    for i in xrange(K):
        if dic[cur] < 0 or flg:
            dic[cur] = i
            cur = A[cur]
        else:
            tmp = K - (dic[cur] - 1)
            tmp %= i - dic[cur]
            i = K - tmp
            flg = True
    print cur

if __name__ == '__main__':
    main()

File: output/ChatGPT/codenet/C++/Java/s877835664.java
import java.util.*;

public class s877835664 {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long N = scanner.nextLong();
        long K = scanner.nextLong();
        
        List<Long> A = new ArrayList<>();
        A.add(0L); // dummy element
        for (int i = 1; i < N; i++) {
            A.add(scanner.nextLong());
        }
        
        long cur = 1;
        List<Long> dic = new ArrayList<>(Collections.nCopies((int) (N + 1), -1L));
        boolean flg = false;
        for (long i = 1; i <= K; i++) {
            if (dic.get((int) cur) < 0 || flg) {
                dic.set((int) cur, i);
                cur = A.get((int) cur);
            } else if (!flg) {
                long tmp = K - (dic.get((int) cur) - 1);
                tmp %= i - dic.get((int) cur);
                i = K - tmp;
                flg = true;
            }
        }
        
        System.out.println(cur);
    }
}